kv.go

```diff
@@ -0,0 +1,387 @@
+type KeyValueUpdate func(v *KeyValueEntry)
+
+// Errors
+var (
+	ErrBucketNameRequired = errors.New("nats: bucket name is required")
+	ErrInvalidBucketName  = errors.New("nats: invalid bucket name")
+	ErrBucketNotFound     = errors.New("nats: bucket not found")
+	ErrBadBucket          = errors.New("nats: bucket not valid key-value store")
+)
+
+const (
+	kvBucketNameTmpl  = "KV_%s"
+	kvSubjectsTmpl    = "$KV.%s.>"
+	kvSubjectsPreTmpl = "$KV.%s."
+	kvNoPending       = "0"
+)
+
+// KeyValue will lookup and bind to an existing KeyValue store.
+func (nc *Conn) KeyValue(bucket string, opts ...JSOpt) (KeyValue, error) {
+	if bucket == _EMPTY_ {
+		return nil, ErrBucketNameRequired
+	}
+	if strings.Contains(bucket, ".") {
+		return nil, ErrInvalidBucketName
+	}
+	jsc, err := nc.JetStream(opts...)
+	if err != nil {
+		return nil, err
+	}
+	stream := fmt.Sprintf(kvBucketNameTmpl, bucket)
+	si, err := jsc.StreamInfo(stream)
+	if err != nil {
+		if err == ErrStreamNotFound {
+			err = ErrBucketNotFound
+		}
+		return nil, err
+	}
+	// Do some quick sanity checks that this is a correctly formed stream for KV.
+	// Max msgs per subject should be > 0.
+	if si.Config.MaxMsgsPerSubject < 1 {
+		return nil, ErrBadBucket
+	}
+
+	kv := &kvs{
+		name:   bucket,
+		stream: stream,
+		pre:    fmt.Sprintf(kvSubjectsPreTmpl, bucket),
+		nc:     nc,
+		js:     jsc.(*js),
+	}
+	return kv, nil
+}
+
+// AddKeyValue will create a KeyValue store with the following configuration.
+func (nc *Conn) AddKeyValue(cfg *KeyValueConfig, opts ...JSOpt) (KeyValue, error) {
+	if cfg == nil || cfg.Bucket == _EMPTY_ {
+		return nil, ErrBucketNameRequired
+	}
+	if strings.Contains(cfg.Bucket, ".") {
+		return nil, ErrInvalidBucketName
+	}
+
+	jsc, err := nc.JetStream(opts...)
+	if err != nil {
+		return nil, err
+	}
+	if _, err = jsc.AccountInfo(); err != nil {
+		return nil, err
+	}
+
+	// Default to 1 for history.
+	history := int64(1)
+	if cfg.History > 0 {
+		history = int64(cfg.History)
+	}
+	replicas := cfg.Replicas
+	if replicas == 0 {
+		replicas = 1
+	}
+
+	scfg := &StreamConfig{
+		Name:              fmt.Sprintf(kvBucketNameTmpl, cfg.Bucket),
+		Description:       cfg.Description,
+		Subjects:          []string{fmt.Sprintf(kvSubjectsTmpl, cfg.Bucket)},
+		MaxMsgsPerSubject: history,
+		MaxMsgs:           cfg.MaxMsgs,
+		MaxBytes:          cfg.MaxBytes,
+		MaxAge:            cfg.MaxAge,
+		MaxMsgSize:        cfg.MaxValueSize,
+		Storage:           cfg.Storage,
+		Replicas:          replicas,
+	}
+
+	if _, err := jsc.AddStream(scfg); err != nil {
+		return nil, err
+	}
+
+	kv := &kvs{
+		name:   cfg.Bucket,
+		stream: scfg.Name,
+		pre:    fmt.Sprintf(kvSubjectsPreTmpl, cfg.Bucket),
+		nc:     nc,
+		js:     jsc.(*js),
+	}
+	return kv, nil
+}
+
+type kvs struct {
+	name   string
+	stream string
+	pre    string
+	nc     *Conn
+	js     *js
+}
+
+// Get returns the latest value for the key.
+func (kv *kvs) Get(key string) (value []byte, revision uint64, err error) {
```

Comment:
Should return KeyValueEntry, languages can add language specific typed gets like this would be GetValue or GetBytes or something rather.  But important that we give access to the KeyValueEntry else you have no idea about historical state, delete operations etc

Reply:
I am not sure, be good to discuss.

Reply:
We did discuss these things at length during the ADR desgin process.  You gave a lot of feedback on the various data types of the `Item` property etc.

Reply:
Yeah, I found little utility for getting the entry except for the ability to know if a key was deleted. But it makes sense to expose the `getEntry` call because you get the entry anyway, it is the entire message. I in turn made `get`  just return the value (or null when key is missing or deleted) because to me that's the most common use. Probably should update the ADR